Vector RecordVideo

sparisi opened the pull request:
# Description

Vector RecordVideo wrapper. Related to [this issue](https://github.com/Farama-Foundation/Gymnasium/issues/1414#issuecomment-3084989093).
Docs of the new class has an example.

I copied the basic RecordVideo wrapper and changed `_capture_frame` to process tuples of frames (from VectorEnv). Then, there are two helper functions `_get_concat_frame_shape` and `_concat_frames` that concatenate these frames into a "big frame" depending on the aspect ratio passed by the user at `__init__`. 
It would be nice to have default `aspect_ratio=None` and use the same logic of Vector HumanRendering, but I am not sure how it works. 
I also made new tests that are the same as basic RecordVideo (except the one with RenderCollection since there is no Vector version of it).

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files` (see `CONTRIBUTING.md` instructions to set it up)
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

sparisi commented on gymnasium/wrappers/vector/rendering.py:
```diff
@@ -182,3 +187,278 @@ def close(self):
+    ) -> tuple[ObsType, dict[str, Any]]:
+        """Reset the environment and eventually starts a new recording."""
+        obs, info = super().reset(seed=seed, options=options)
+        self.episode_id += 1
```
I copied this from [regular RecordVideo.](https://github.com/Farama-Foundation/Gymnasium/blob/58c6b7b49da20418b21df632866eb215f8c26876/gymnasium/wrappers/rendering.py#L340) 
We can add your suggestions, but what do you mean exactly? How can I make an environment with auto-resets?

sparisi commented on gymnasium/wrappers/vector/rendering.py:
```diff
@@ -182,3 +187,278 @@ def close(self):
+            self._capture_frame()
+
+            if len(self.recorded_frames) > self.video_length:
+                self.stop_recording()
```
What is env_id?
(Also copied [from here](https://github.com/Farama-Foundation/Gymnasium/blob/58c6b7b49da20418b21df632866eb215f8c26876/gymnasium/wrappers/rendering.py#L371))

sparisi commented on gymnasium/wrappers/vector/rendering.py:
```diff
@@ -182,3 +187,278 @@ def close(self):
+
+        if isinstance(frame, list):
+            if len(frame) == 0:  # render was called
+                return
```
I just followed the [regular implementation.](https://github.com/Farama-Foundation/Gymnasium/blob/58c6b7b49da20418b21df632866eb215f8c26876/gymnasium/wrappers/rendering.py#L327)

pseudo-rnd-thoughts replied:
Could you add an exception (or at least a warning) to both wrappers, as we shouldn't get an empty rendering because it means that we would skip a frame from an episode

pseudo-rnd-thoughts reviewed (changes requested):
With autoreset, we need to detect episode resets in `step` using `terminated[0] or truncated[0]` on if the first sub-environment has ends / been autoresetted

Could you add a warning to the constructor that if the vector environment metadata autoreset-mode is SameStep, then "Recorded episodes will not contain the last frame of the episode"

sparisi commented on gymnasium/wrappers/vector/rendering.py:
```diff
@@ -182,3 +187,278 @@ def close(self):
+        """Steps through the environment using action, recording observations if :attr:`self.recording`."""
+        obs, rew, terminated, truncated, info = self.env.step(action)
+        self.step_id += 1
+
```
1. Wouldn't episode_id increase by twice with proper resets? Once by the new like you suggested to add, once by the +1 in reset().
2. Should we check all envs? 
self.episode_id += np.logical_or(truncated, terminated).sum()

sparisi commented on gymnasium/wrappers/vector/rendering.py:
```diff
@@ -182,3 +187,278 @@ def close(self):
+
+            if len(self.recorded_frames) > self.video_length:
+                self.stop_recording()
+
```
You mean episode_id += 1? It is already increase at the beginnin of the function. (And from the tests, it seems that videos are numbered correctly with episode trigger).

pseudo-rnd-thoughts reviewed (changes requested):
Thanks for reminding me @sparisi 

Looking at the code again, quite a few updates are necessary with the changes for moving to vector environments. 

The primary problem still arises from autoreset as for vector environments (unless using disable autoreset) then users generally never call `reset` after the initial call. 
Therefore, we need to add code to handle the three autoreset modes in `step` and in the tests as they are currently a direct copy from the standard environment that does use frequent resets.

sparisi commented:
@pseudo-rnd-thoughts Sounds good, you can ping me when things are ready so I can get back to this PR 👍

sparisi commented:
> @sparisi I think I've updated the implementation to account for autoreset however I haven't tested this practically to check that the video outputs are as expected

Hey, sorry for the late update.
I fixed a small bug in the doc and tested all reset modes. "Disabled" raises 

```
  File "C:\Users\Simone\Documents\GitHub\Gymnasium\gymnasium\vector\sync_vector_env.py", line 268, in step
    assert not self._autoreset_envs[i], f"{self._autoreset_envs=}"
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: self._autoreset_envs=array([False, False, False, False, False, False, False, False,  True,
       False])
```

Is this expected?

```
envs = gym.make_vec(
  "LunarLander-v3", 
  num_envs=10,
  render_mode="rgb_array", 
  vectorization_mode=gym.VectorizeMode.SYNC, 
  vector_kwargs={"autoreset_mode": "Disabled"},
)
envs = RecordVideo(
    envs,
    video_folder="save_videos_5envs",
    video_aspect_ratio=(1,1),
    episode_trigger=lambda t: t % 5 == 0,
)
_ = envs.reset(seed=123)
_ = envs.action_space.seed(123)
for i in range(400):
    actions = envs.action_space.sample()
    _ = envs.step(actions)
envs.close()
```
